### ISSUE TITLE
BUG: Slow function call in `scipy.optimize.minimize` with `L-BFGS-B` since `scipy>=1.15.0`  on Ubuntu

### ISSUE THREAD

**Issue by nabenabe0928** [scipy/scipy#23191]
### Describe your issue.

This issue relates to:
- https://github.com/scipy/scipy/issues/22438

Although the previous issue discusses the cause of the slowdown, the conclusion there seems not universal based on my observation.
I basically observed the slowdown only from the 38-th to 51-th calls when using the example code below and this phenomenon reproduces every time I try, which seems very deterministic and does not explain the cause by multi-threading.

I suspect there are some other reasons why it slows down.

The only difference in the benchmarking was SciPy, which I used `scipy==1.14.1` and `scipy==1.15.3`.

> [!NOTE]
> Although the slowdown became milder with the suggestion above, which is to limit the thread count by exporting `OPENBLAS_NUM_THREADS=1`, I think there are some additional causes of the slowdown.

> [!IMPORTANT]
> In my environment, `with threadpool_limits(limits=1, user_api="blas")` did not remove the slowdown with `threadpoolctl==3.6.0`.

### Reproducing Code Example

```python
I used the following code to observe the slow down:


import math
import numpy as np
import scipy
import torch
import time


n_points = 128
seed = 5
torch.manual_seed(seed)
np.random.seed(seed)
rand_tensor = torch.rand(n_points, n_points, n_points)
counter = 0


def func(x):
    global counter
    counter += 1
    start = time.time()
    X_tensor = torch.from_numpy(x).requires_grad_(True)
    diff = rand_tensor - X_tensor
    loss = (diff @ diff).sum()
    grad = torch.autograd.grad(loss, X_tensor)[0].reshape(-1).numpy(force=True)
    print(f"Func call {counter:0>2}:", time.time() - start)
    return loss.item(), grad


x0 = np.random.random(n_points)
start = time.time()
res = scipy.optimize.minimize(
    func,
    x0.copy(),
    jac=True,
    method="l-bfgs-b",
    options={"gtol": 1e-2},
)
print(time.time() - start, scipy.__version__)



> [!NOTE]
> This issue goes away when using the following code:

from contextlib import contextmanager
import os


@contextmanager
def single_blas_thread() -> None:
    old_val = os.environ.get("OPENBLAS_NUM_THREADS")
    os.environ["OPENBLAS_NUM_THREADS"] = "1"
    try:
        yield
    finally:
        if old_val is None:
            os.environ.pop("OPENBLAS_NUM_THREADS", None)
        else:
            os.environ["OPENBLAS_NUM_THREADS"] = old_val


with single_blas_thread():
    res = scipy.optimize.minimize(...)
```

### Error message

```shell
No error message.
```

### SciPy/NumPy/Python version and system information

```shell
The shared environment is as follows:

- OS: Ubuntu 20.04
- Python Version: 3.12.11
- Python Package Dependencies:


filelock==3.13.1
fsspec==2024.6.1
Jinja2==3.1.4
MarkupSafe==2.1.5
mpmath==1.3.0
networkx==3.3
numpy==2.2.6
setuptools==70.2.0
sympy==1.13.3
torch==2.7.1+cpu
typing_extensions==4.12.2
```

**Comment by nabenabe0928** [comment]
Another finding is that it seems the routine becomes significantly slower when the gradient mean becomes huge.

<details>

<summary>Log with Mean Gradient and Func Call Durations</summary>

```
Func call 01: 2.963e+01[ms]
Grad.: 4.296e+04
Func call 02: 2.784e+01[ms]
Grad.: -3.254e+05
Func call 03: 2.957e+01[ms]
Grad.: -5.516e+02
Func call 04: 2.865e+01[ms]
Grad.: -5.516e+02
Func call 05: 3.268e+01[ms]
Grad.: -5.516e+02
Func call 06: 2.446e+01[ms]
Grad.: -5.516e+02
Func call 07: 2.751e+01[ms]
Grad.: -5.516e+02
Func call 08: 2.448e+01[ms]
Grad.: -5.516e+02
Func call 09: 2.855e+01[ms]
Grad.: -5.516e+02
Func call 10: 2.666e+01[ms]
Grad.: -5.516e+02
Func call 11: 2.757e+01[ms]
Grad.: -5.516e+02
Func call 12: 2.262e+01[ms]
Grad.: -5.516e+02
Func call 13: 2.397e+01[ms]
Grad.: -5.516e+02
Func call 14: 2.346e+01[ms]
Grad.: -5.516e+02
Func call 15: 3.313e+01[ms]
Grad.: -5.516e+02
Func call 16: 2.760e+01[ms]
Grad.: -5.516e+02
Func call 17: 2.343e+01[ms]
Grad.: -5.516e+02
Func call 18: 2.259e+01[ms]
Grad.: -5.516e+02
Func call 19: 2.789e+01[ms]
Grad.: -5.516e+02
Func call 20: 2.451e+01[ms]
Grad.: -5.515e+02
Func call 21: 2.788e+01[ms]
Grad.: -5.514e+02
Func call 22: 2.718e+01[ms]
Grad.: -5.516e+02
Func call 23: 2.473e+01[ms]
Grad.: -5.523e+02
Func call 24: 2.634e+01[ms]
Grad.: -5.554e+02
Func call 25: 2.669e+01[ms]
Grad.: -5.674e+02
Func call 26: 2.712e+01[ms]
Grad.: -6.156e+02
Func call 27: 2.710e+01[ms]
Grad.: -8.083e+02
Func call 28: 2.912e+01[ms]
Grad.: -1.579e+03
Func call 29: 2.818e+01[ms]
Grad.: -4.663e+03
Func call 30: 2.768e+01[ms]
Grad.: -1.700e+04
Func call 31: 2.455e+01[ms]
Grad.: -6.634e+04
Func call 32: 2.373e+01[ms]
Grad.: -2.637e+05
Func call 33: 2.590e+01[ms]
Grad.: -1.053e+06
Func call 34: 2.537e+01[ms]
Grad.: -4.211e+06
Func call 35: 2.880e+01[ms]
Grad.: -1.684e+07
Func call 36: 3.244e+01[ms]
Grad.: -1.275e+08
Func call 37: 4.237e+01[ms]
Grad.: -2.744e+08
Func call 38: 1.226e+02[ms]
Grad.: -5.320e+08
Func call 39: 1.233e+02[ms]
Grad.: -9.364e+08
Func call 40: 1.115e+02[ms]
Grad.: -1.598e+09
Func call 41: 1.248e+02[ms]
Grad.: -2.665e+09
Func call 42: 1.101e+02[ms]
Grad.: -4.393e+09
Func call 43: 1.109e+02[ms]
Grad.: -7.189e+09
Func call 44: 1.127e+02[ms]
Grad.: -1.171e+10
Func call 45: 1.312e+02[ms]
Grad.: -1.903e+10
Func call 46: 1.356e+02[ms]
Grad.: -3.087e+10
Func call 47: 1.226e+02[ms]
Grad.: -5.004e+10
Func call 48: 1.194e+02[ms]
Grad.: -8.105e+10
Func call 49: 1.228e+02[ms]
Grad.: -1.312e+11
Func call 50: 1.116e+02[ms]
Grad.: -2.125e+11
Func call 51: 1.214e+02[ms]
Grad.: -3.440e+11
Func call 52: 2.217e+01[ms]
Grad.: 1.443e+18
Func call 53: 2.725e+01[ms]
Grad.: 9.689e+01
Func call 54: 2.800e+01[ms]
Grad.: 1.264e+02
Func call 55: 2.736e+01[ms]
Grad.: 2.543e+02
Func call 56: 2.921e+01[ms]
Grad.: 1.264e+02
Func call 57: 2.990e+01[ms]
Grad.: 1.963e+02
Func call 58: 3.276e+01[ms]
Grad.: 1.264e+02
Func call 59: 2.926e+01[ms]
Grad.: 1.704e+02
Func call 60: 2.805e+01[ms]
Grad.: 1.264e+02
Func call 61: 2.924e+01[ms]
Grad.: 1.006e+02
Func call 62: 2.748e+01[ms]
Grad.: 1.710e+02
Func call 63: 2.986e+01[ms]
Grad.: 1.006e+02
Func call 64: 2.837e+01[ms]
Grad.: 1.317e+02
Func call 65: 2.967e+01[ms]
Grad.: 1.006e+02
Func call 66: 2.585e+01[ms]
Grad.: 7.673e+01
Func call 67: 2.619e+01[ms]
Grad.: 1.006e+02
Func call 68: 3.057e+01[ms]
Grad.: 1.682e+02
Func call 69: 2.785e+01[ms]
Grad.: 1.006e+02
Func call 70: 3.006e+01[ms]
Grad.: 1.001e+02
Func call 71: 2.831e+01[ms]
Grad.: 1.006e+02
Func call 72: 2.997e+01[ms]
Grad.: 1.001e+02
Func call 73: 2.735e+01[ms]
Grad.: 1.006e+02
```



</details>

**Comment by ilayn** [comment]
Thank you @nabenabe0928 for the new issue. This is indeed very confusing for me since I can only replicate things on this Ubuntu virtual machine and nowhere else (within my reach). 

Do you have any means to test the code on other machines of any sort? I just don't want to triangulate things to Ubuntu if it is a false diagnosis. I just want to make sure that we are all having this issue with Ubuntu based distros.

**Comment by kAIto47802** [comment]
@ilayn 
I experienced the same slowdown while I was using Arch Linux, which is a completely different distribution from Ubuntu.
I used Python 3.11.0 and tested both SciPy versions 1.15.0 and 1.16.0.

**Comment by rgommers** [comment]
Both this reproducer and gh-22438 use PyTorch in addition to NumPy/SciPy. That causes both MKL and OpenBLAS to be pulled in (worse actually, you'll get two OpenBLAS builds), and they may both be messing with threading somehow. A few questions:

1. Does the fix/workaround at https://github.com/numpy/numpy/issues/29352#issuecomment-3059188352 help?
2. Is it possible to reproduce the slowdown for anyone without `import torch`?
3. Does it depend on import order (if you do `import torch` first, does the slowdown still happen)?
4. On a machine where the slowdown reproduces, can someone run it after adding in `pprint(threadpool.threadpool_info())` (as documented [here](https://github.com/joblib/threadpoolctl?tab=readme-ov-file#python-runtime-programmatic-introspection)) before and after each import?

**Comment by isuruf** [comment]
~~On my AMD cpu with 2 threads per core and 8 threads, MKL uses max `8` threads and OpenMP` max threads is `16`. PyTorch sets the number of OpenMP threads to the maximum of MKL threads. See https://github.com/pytorch/pytorch/blob/55c289d5c104c4959cc125c0fb4fb50c9fc71102/aten/src/ATen/ParallelOpenMP.cpp#L33-L38~~

~~When I change the max CPU threads to `16`, I get the performant path. `1` is also better than `8`.~~

**Comment by isuruf** [comment]
Hmm, scratch that `MKL` has two functions `mkl_domain_get_max_threads()` gives 8 and `mkl_get_max_threads()` gives 16
